MoltenVK/MoltenVK/GPUObjects/MVKImage.mm

```diff
@@ -561,6 +561,142 @@
+		MTLPixelFormat mtlPixFmt = mtlTex.pixelFormat;
+		bool isPVRTC = pixFmts->isPVRTCFormat(mtlPixFmt);
+
+		uint32_t texelsWidth = imgRgn.memoryRowLength ? imgRgn.memoryRowLength : imgRgn.imageExtent.width;
+		uint32_t texelsHeight = imgRgn.memoryImageHeight ? imgRgn.memoryImageHeight : imgRgn.imageExtent.height;
+		uint32_t texelsDepth = imgRgn.imageExtent.depth;
+		size_t rowPitch = pixFmts->getBytesPerRow(mtlPixFmt, texelsWidth);
+		size_t depthPitch = pixFmts->getBytesPerLayer(mtlPixFmt, rowPitch, texelsHeight);
+		size_t arrayPitch = depthPitch * texelsDepth;
+
+		MTLRegion mtlRgn;
+		mtlRgn.origin = mvkMTLOriginFromVkOffset3D(imgRgn.imageOffset);
+		mtlRgn.size = mvkMTLSizeFromVkExtent3D(imgRgn.imageExtent);
+
+		for (uint32_t imgLyrIdx = 0; imgLyrIdx < imgSubRez.layerCount; imgLyrIdx++) {
+			void* pImgBytes = (void*)((uintptr_t)imgRgn.pHostPointer + (arrayPitch * imgLyrIdx));
+			if (toImage) {
+				VkSubresourceLayout imgLayout = { 0, 0, rowPitch, arrayPitch, depthPitch};
+#if MVK_MACOS
+				// Compressed content cannot be directly uploaded to a compressed 3D texture.
+				// But we can upload the decompressed image data.
+				std::unique_ptr<char[]> decompBuffer;
+				if (_is3DCompressed) {
+					std::unique_ptr<MVKCodec> codec = mvkCreateCodec(pixFmts->getVkFormat(mtlPixFmt));
+					if ( !codec ) { return reportError(VK_ERROR_FORMAT_NOT_SUPPORTED, "A 3D texture used a compressed format that MoltenVK does not yet support."); }
+					VkSubresourceLayout linearLayout = {};
+					linearLayout.rowPitch = 4 * imgRgn.imageExtent.width;
+					linearLayout.depthPitch = linearLayout.rowPitch * imgRgn.imageExtent.height;
+					linearLayout.size = linearLayout.depthPitch * imgRgn.imageExtent.depth;
+					decompBuffer = std::unique_ptr<char[]>(new char[linearLayout.size]);
+					codec->decompress(decompBuffer.get(), pImgBytes, linearLayout, imgLayout, imgRgn.imageExtent);
+					pImgBytes = decompBuffer.get();
+					imgLayout = linearLayout;
+				}
+#endif
+				[mtlTex replaceRegion: mtlRgn
+						  mipmapLevel: imgSubRez.mipLevel
+								slice: imgSubRez.baseArrayLayer + imgLyrIdx
+							withBytes: pImgBytes
+						  bytesPerRow: (isPVRTC || is1D) ? 0 : imgLayout.rowPitch
+						bytesPerImage: (isPVRTC || !is3D) ? 0 : imgLayout.depthPitch];
+			} else {
+				[mtlTex getBytes: pImgBytes
+					 bytesPerRow: (isPVRTC || is1D) ? 0 : rowPitch
+				   bytesPerImage: (isPVRTC || !is3D) ? 0 : depthPitch
+					  fromRegion: mtlRgn
+					 mipmapLevel: imgSubRez.mipLevel
+						   slice: imgSubRez.baseArrayLayer + imgLyrIdx];
+			}
+		}
+	}
+	return VK_SUCCESS;
+}
+
+// Create concrete implementations of the variations of the copyContent() template function.
+// This is required since the template is called from outside this file (compilation unit).
+template VkResult MVKImage::copyContent(const VkCopyMemoryToImageInfoEXT* pCopyInfo, bool toImage);
+template VkResult MVKImage::copyContent(const VkCopyImageToMemoryInfoEXT* pCopyInfo, bool toImage);
```

Comment:
The design purist in me doesn't like that the `toImage` parameter is necessary here  to avoid duplicating all the setup code, since it will always be `true` in the first case and `false` in the second. Maybe if the parts that performed the actual copy (inside the `if (toImage)` blocks above) were factored out into separate methods, and then the right method would get called in `copyContent()` based on the type of the copy info...

Reply:
Great idea. I'm also a fan of design elegance, and I've implemented what you suggested.